### PR TITLE
Paquet and JSON readers use Arc<Expression> to avoid deep copies

### PR DESCRIPTION
**Pull request by scovich**
Today, the engine parquet/json file handler APIs take an `Expression` arg for predicate pushdown. They cannot take a reference, because the iterator they return will likely depend on (but outlive) that reference. Worse, they need to do it for every file the query reads. Unfortunately, data skipping predicates can be arbitrarily large, and thus annoying/expensive to copy so much. We already use Arc to protect schemas (some of the time, at least), and we can start using Arc to protect expressions as well.

### CODE REVIEW

File: kernel/src/scan/mod.rs [hunk 1]
```diff
@@ -467,7 +464,7 @@ fn transform_to_logical_internal(
             .get_expression_handler()
             .get_evaluator(
                 read_schema,
-                read_expression.clone(),
```
Comment (scovich) [inline kernel/src/scan/mod.rs:470]: One copy saved

File: kernel/src/snapshot.rs [hunk 2]
```diff
@@ -81,12 +80,14 @@ impl LogSegment {
         let schema = get_log_schema().project(&[PROTOCOL_NAME, METADATA_NAME])?;
         // filter out log files that do not contain metadata or protocol information
         use Expression as Expr;
-        let meta_predicate = Some(Expr::or(
-            Expr::not(Expr::is_null(Expr::column("metaData.id"))),
-            Expr::not(Expr::is_null(Expr::column("protocol.minReaderVersion"))),
-        ));
+        lazy_static::lazy_static!(
```
Comment (scovich) [inline kernel/src/snapshot.rs:83]: Meta-predicates don't change, so we can define them as lazy static arc. And then just shallow-clone the arc as needed.

(again below)

File: kernel/src/scan/mod.rs [hunk 3]
```diff
@@ -74,18 +74,15 @@ impl ScanBuilder {
...
     /// Optionally provide an [`Expression`] to filter rows. See [`ScanBuilder::with_predicate`] for
     /// details. If `predicate_opt` is `None`, this is a no-op.
-    pub fn with_predicate_opt(self, predicate_opt: Option<Expression>) -> Self {
-        match predicate_opt {
-            Some(predicate) => self.with_predicate(predicate),
-            None => self,
-        }
+    pub fn with_predicate_opt(mut self, predicate: Option<Expression>) -> Self {
+        self.predicate = predicate.map(Arc::new);
+        self
```
[the remark this answers is not in the document]
Comment (scovich) [reply kernel/src/scan/mod.rs:85]: That said, we should probably remove one of the overloads and just be consistent at all call sites?

File: kernel/src/snapshot.rs (hunk 2 above)
Comment (hntd187) [reply kernel/src/snapshot.rs:83]: We should probably use `OnceLock` now that it's in std.

https://doc.rust-lang.org/stable/std/sync/struct.OnceLock.html

File: kernel/src/scan/data_skipping.rs [hunk 4]
```diff
@@ -188,11 +189,7 @@ impl DataSkippingFilter {
             static ref FILTER_EXPR: Expr = Expr::column("predicate").distinct(Expr::literal(false));
         );
 
-        let predicate = match predicate {
-            Some(predicate) => predicate,
-            None => return None,
-        };
-
+        let predicate = predicate.as_deref()?;
```
[the remark this answers is not in the document]
Comment (hntd187) [reply kernel/src/scan/data_skipping.rs:192]: Because Try isn't implemented on `&Option` as_deref converts `&Option<T>` to `Option<&T>` https://doc.rust-lang.org/std/option/enum.Option.html#method.as_deref

File: kernel/src/scan/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicklan) [reply kernel/src/scan/mod.rs:85]: Yeah, I think we should just have `with_predicate_opt`, or maybe just rename `with_predicate_opt` to `with_predicate` and get rid of the one that doesn't take an option.

**Review by nicklan — COMMENTED** [review]
given that this matches what arrow does with their various `XRef` types, I'm inclined to think yeah this is a good idea.

File: kernel/src/scan/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (scovich) [reply kernel/src/scan/mod.rs:85]: This is weird... I'm not seeing where predicates get assigned at all in our code? The FFI layer has a call to `with_predicate`, but the only other callers are in tests/read.rs. Do we know what e.g. delta-rs uses?

[the remark this answers is not in the document]
Comment (nicklan) [reply kernel/src/scan/mod.rs:85]: delta-rs isn't at the point of doing scans with kernel yet, so I think only duckdb is exercising this so far

[the remark this answers is not in the document]
Comment (scovich) [reply kernel/src/scan/mod.rs:81]: Fixed.

File: kernel/src/engine/sync/mod.rs [hunk 5]
```diff
@@ -60,12 +60,12 @@ impl Engine for SyncEngine {
 fn read_files<F, I>(
     files: &[FileMeta],
     schema: SchemaRef,
-    predicate: Option<Expression>,
+    predicate: Option<ExpressionRef>,
     mut try_create_from_file: F,
 ) -> DeltaResult<FileDataReadResultIterator>
 where
     I: Iterator<Item = DeltaResult<ArrowEngineData>> + Send + 'static,
-    F: FnMut(File, SchemaRef, ArrowSchemaRef, Option<&Expression>) -> DeltaResult<I>
```
Comment (zachschuermann) [inline kernel/src/engine/sync/mod.rs:68]: checking my understanding: in places where we previously took `&Expression` we are switching to Arc to allow for the expression 'sharing' we want? to have this expression share among many places via Arc? if we left as &Expression we would have to have an owned one somewhere which we are handing out references to?

File: kernel/src/scan/mod.rs [hunk 6]
```diff
@@ -70,24 +70,17 @@ impl ScanBuilder {
...
+    /// 4` to return a subset of the rows in the scan which satisfy the filter. If `predicate_opt`
+    /// is `None`, this is a no-op.
     ///
-    /// Can be used to filter the rows in a scan. For example, using the predicate
-    /// `x < 4` to return a subset of the rows in the scan which satisfy the filter.
-    pub fn with_predicate(mut self, predicate: Expression) -> Self {
-        self.predicate = Some(predicate);
+    /// NOTE: The filtering is best-effort and can produce false positives (rows that should should
+    /// have been filtered out but were kept).
+    pub fn with_predicate(mut self, predicate: impl Into<Option<Expression>>) -> Self {
```
Comment (zachschuermann) [inline kernel/src/scan/mod.rs:79]: question: how did you decide on `Expression` vs `Arc<Expression>` as method parameter?

File: kernel/src/scan/mod.rs (hunk 1 above)
Comment (zachschuermann) [reply kernel/src/scan/mod.rs:509]: nice!

**Review by zachschuermann — APPROVED** [review]
looks great just added a couple questions of my own :)

File: kernel/src/engine/sync/mod.rs (hunk 5 above)
Comment (scovich) [reply kernel/src/engine/sync/mod.rs:68]: That is a (small) part of it. The main reason is that iterators tend to outlive the method invocation that creates them and that makes it really difficult to track lifetimes of a borrowed reference the iterator needs to consume. And it's outright impossible if there's any async code lurking underneath.

File: kernel/src/scan/mod.rs (hunk 6 above)
Comment (scovich) [reply kernel/src/scan/mod.rs:79]: No strong reason, just convenience.

This method turns out to be a main entry point from the outside world. Mostly unit tests and examples, which anyway create the expressions from scratch. Including FFI where we _have_ to create a new instance because native engines can't instantiate an Arc even if they wanted to.

Do you think there's a non-trivial risk that connector writers might get frustrated because they have to repeatedly deep-clone an `ExpressionRef` they already had lying around for some reason? Maybe dashboard queries or something?

Comment (zachschuermann) [reply kernel/src/scan/mod.rs:79]: hm I don't have enough intuition/information to answer that - seems fine for now given that it's our preference to pass an owned `Expression` (from usability in UT etc perspective) but definitely wanted to highlight the possible risk you just mentioned: that the engine has an `Arc<Expression>` laying around and wants to hand it out instead of forcing an owned Expression be passed into this method.

TLDR it gives _us_ (kernel) the flexibility to share the expression after this point but not above it (engine sharing it). right? perhaps we just ship this and change if anyone comes asking :) just noticed this and wanted to ask!

Comment (nicklan) [reply kernel/src/scan/mod.rs:79]: Yeah, seems this could be `impl Into<Option<ExpressionRef>>`?

**Review by nicklan — APPROVED** [review]
lgtm, one question that zach had too
